runtime/src/main/java/org/corfudb/runtime/object/transactions/TransactionalContext.java

```diff
@@ -86,6 +110,7 @@ public static AbstractTransactionalContext newContext(AbstractTransactionalConte
      */
     public static AbstractTransactionalContext removeContext() {
         AbstractTransactionalContext r = getTransactionStack().pollFirst();
+        contextSet.remove(r);
```

Comment:
I would add something like:

     public static AbstractTransactionalContext removeContext() {
Deque<AbstractTransactionalContext> stack = getTransactionStack();
if (stack.isEmpty()) {
    threadTransactionStack.set(null);
}

Otherwise if the current thread is polled and it is returned back to the pool, you are doing a small memory leak.